module/VuFind/src/VuFind/RateLimiter/RateLimiterManager.php

```diff
@@ -318,4 +343,68 @@ protected function isCrawlerRequest(MvcEvent $event): bool
         $crawlerDetect = new \Jaybizzle\CrawlerDetect\CrawlerDetect();
         return $crawlerDetect->isCrawler($agent);
     }
+
+    /**
+     * Check whether the RateLimiter policy for this event uses Turnstile.
+     *
+     * @param MvcEvent $event Request event
+     *
+     * @return bool
+     */
+    public function checkPolicyUsesTurnstile($event)
+    {
+        $policyId = $this->getPolicyIdForEvent($event);
+        if ($this->config['Policies'][$policyId]['turnstileRateLimiterSettings'] ?? false) {
+            return $policyId;
+        }
+        return false;
+    }
+
+    /**
+     * Check for a prior, cached result from Turnstile under this client IP and policy.
+     *
+     * @param string $policyId The policy ID
+     * @param string $clientIp The client IP
+     *
+     * @return ?bool Null if there is no prior result, or if Turnstile is disabled;
+     *               otherwise a boolean representing the Turnstile result.
+     */
+    protected function checkPriorTurnstileResult($policyId, $clientIp)
+    {
+        if (!($this->config['Policies'][$policyId]['turnstileRateLimiterSettings'] ?? false)) {
+            return null;
+        }
+        $cacheKey = $this->getTurnstileCacheKey($policyId, $clientIp);
+        return $this->turnstileCache->getItem($cacheKey);
+    }
+
+    /**
+     * Store a Turnstile result for this client IP and policy.
+     *
+     * @param string $policyId The policy ID
+     * @param string $clientIp The client IP
+     * @param bool   $success  The result to store
+     *
+     * @return void
+     */
+    public function setTurnstileResult($policyId, $clientIp, $success)
+    {
+        $cacheKey = $this->getTurnstileCacheKey($policyId, $clientIp);
+        $this->turnstileCache->setItem($cacheKey, $success);
+    }
+
+    /**
+     * Generate a key for the Turnstile cache.
+     *
+     * @param string $policyId The policy ID
+     * @param string $clientIp The client IP
+     *
+     * @return string The cache key
+     */
+    protected function getTurnstileCacheKey($policyId, $clientIp)
+    {
+        $key = $policyId . '--' . $clientIp;
+        $key = str_replace('.', '-', $key);
```

Comment:
The (temporary) cache used now complained about the dots in the IP address.  This is possibly moot once we have a real cache, depending what its rules are for keys.

Reply:
Would some kind of hash be a more foolproof way of creating predictable keys?